[doc] Add documentation guide for MPI on Ray. (#41626)
ray-project/ray: doc/source/ray-core/mpi-cluster.rst

.. _ray-mpi-guide:

Run MPI cluster on Ray (Experimental)
=====================================

.. note::

    This feature is experimental and the API may change in future releases.

`MPI <https://www.open-mpi.org/>`_ is a popular framework for distributed

> computation. MPI programs are supported to run on Ray clusters. This guide

Reviewer: ```suggestion
computation. Ray clusters support the running of MPI programs. This guide
```

describes how to run MPI programs on Ray clusters.

Requirements
------------

`mpi4py <https://mpi4py.readthedocs.io/en/stable/>`_ is required to run MPI

> programs on Ray clusters. Check the documentation for installation instructions.

Reviewer: ```suggestion
programs on Ray clusters. See the documentation for installation instructions.
```

Get Started
-----------

A typical MPI program consists of a master process and multiple worker
processes. Inside Ray, the master process is a Ray remote function or actor and

> the worker processes are just normal processes. This means the workers can't use

Reviewer: ```suggestion
the worker processes are normal processes. This configuration means the workers can't use
```

Ray APIs, but they can use MPI APIs. The master process can use Ray APIs and MPI
APIs at the same time.

> The support is implemented as a runtime environment plugin. So it has to be

Reviewer: ```suggestion
Ray implements MPI support as a runtime environment plugin. You must
```

> specified in the function or actor decorator. For example:

Reviewer: ```suggestion
specify it in the function or actor decorator. For example:
```

    .. testcode::
      from ray.runtime_env import mpi_init
      @ray.remote(
        runtime_env={
            "mpi": {
                "args": ["-n", "4"],
                "worker_entry": "test_mpi.entry",
            },
        }
      )
      def foo():
        mpi_init()
        pass

The runtime environment plugin has two arguments: `args` and `worker_entry`. The
first one is the arguments passed to `mpirun` and the second one is the entry
function of the worker. The entry function is a normal Python function and it
must be accessible from the worker. In the above example, `foo` is the entry

> function for master process which has rank 0.

Reviewer 2: Yes. Sure.

> `ray.runtime_env.mpi_init` has to be called in the main entry. Otherwise, the

Reviewer: ```suggestion
Call `ray.runtime_env.mpi_init` in the main entry. Otherwise, the
```

> program will hang. This function is to ensure the environment setup is correctly

Reviewer: ```suggestion
program hangs. This function ensures that the environment setup is correct
```

on both the master process and the worker processes.

.. note::
  

>       `ray.runtime_env.mpi_init` is required to be called in the master process before any MPI operations. 

Reviewer: ```suggestion
      You must call `ray.runtime_env.mpi_init` in the master process before any MPI operation. 
```

>       Otherwise, the program will hang.

Reviewer: ```suggestion
      Otherwise, the program hangs.
```

> Actors are also supported. This adds the function for the actor to call MPI, so

Reviewer 3: Can you add explanation when actor is more useful vs task here?

> Actors are also supported. This adds the function for the actor to call MPI, so

Reviewer: ```suggestion
The Ray MPI plugin also supports actors. This code adds the function for an actor to call MPI, so
```

it can utilize MPI to run some distributed computation.

    .. testcode::
      from ray.runtime_env import mpi_init
      @ray.remote(
        runtime_env={
            "mpi": {
                "args": ["-n", "4"],
                "worker_entry": "test_mpi.entry",
            },
        },
        num_gpus=2,
      )
      class Foo:
        def __init__(self):

>           mpi_init()

Reviewer 3: ```suggestion
       # mpi_init has to be called in the beginning of task or actor. Otherwise, it can hang
          mpi_init()
```

          pass

        def bar(self):
          pass

> GPUs are also supported. The number of GPUs can be specified in the actor

Reviewer 3: How does GPU set? Is it each process is assigned to CUDA_VISIBILE_DEVICES?

Reviewer 2: Yes. Each process will have the same device visible envs as the rank 0.

> GPUs are also supported. The number of GPUs can be specified in the actor

Reviewer: ```suggestion
The MPI plugin also supports GPUs. You can specify the number of GPUs in the actor
```

> decorator as above. If it's not set, the program will see all the GPUs of the

Reviewer: ```suggestion
decorator as above. If you don't set it, the program sees all GPUs of the
```

nodes.

> Remote function like `bar` can be called just like normal Ray remote functions.

Reviewer 3: Can you add a simple e2e example here?

> Remote function like `bar` can be called just like normal Ray remote functions.

Reviewer: ```suggestion
You can call remote functions like `bar` in the same way you call normal Ray remote functions.
```

> Inside the bar, it can call MPI APIs to do distributed computation. The

Reviewer: ```suggestion
Inside the bar, it can call MPI APIs to do distributed computation. You
```

> coordination with the workers have to be managed by the user carefully to avoid

Reviewer: ```suggestion
must carefully manage the coordination with the workers to avoid
```
